# FateMappingCode
Collection of code from the [Shendure lab](https://github.com/shendurelab/) fate mapping project using the CRISPR/Cas9 system. This is unsupported code, and stored here as an example of how we did our analysis in the paper, which may or may not be the way you'll want to do your own analysis.

# Abstract

Genome editing of synthetic target arrays for lineage tracing (GESTALT)
Multicellular systems develop from single cells through a lineage, but current lineage tracing approaches scale poorly to whole organisms. Here we use genome editing to progressively introduce and accumulate diverse mutations in a DNA barcode over multiple rounds of cell division. The barcode, an array of CRISPR/Cas9 target sites, records lineage relationships in the patterns of mutations shared between cells. In cell culture and zebrafish, we show that rates and patterns of editing are tunable, and that thousands of lineage-informative barcode alleles can be generated. We find that most cells in adult zebrafish organs derive from relatively few embryonic progenitors. Genome editing of synthetic target arrays for lineage tracing (GESTALT) will help generate large-scale maps of cell lineage in multicellular systems.

The full paper can be found on Science's website here, and our pre-print is available on the [bioRxiv server here](http://biorxiv.org/content/early/2016/05/11/052712).

Data used in the paper is available in the [Dryad repository](doi:10.5061/dryad.478t9), which links out to GEO for sequencing data.
